Core_Method/CoreMethod_FWD/ROS_FWD_Integrator.m

```diff
@@ -205,11 +215,62 @@
                 end
                 
                 % Solve the system
-                K(ioffset+1:ioffset+NVAR) = e\K(ioffset+1:ioffset+NVAR);
-                ISTATUS.Nsol = ISTATUS.Nsol + 1;
+                if ( ~OPTIONS.MatrixFree )
+                    RHS  = K(ioffset+1:ioffset+NVAR);
+                    if issparse(fjac)
+                        K(ioffset+1:ioffset+NVAR) = e.Q * (e.U \ (e.L \ (e.P * (e.R \ RHS))));
+                    else
+                        K(ioffset+1:ioffset+NVAR) = e.U \ (e.L \ RHS(e.p));
+                    end
+                    ISTATUS.Nsol = ISTATUS.Nsol + 1;
+                    
+%                     [ tempK, bicgFlag, ~, iter] = ...
+%                         bicg(e, K(ioffset+1:ioffset+NVAR), ...
+%                              OPTIONS.GMRES_TOL, OPTIONS.GMRES_MaxIt, OPTIONS.GMRES_P);
+%                     ISTATUS.Nsol = ISTATUS.Nsol + 1;
+%                     
+%                     ISTATUS.Njac =  ISTATUS.Njac + iter;
+%                     
+%                     if( bicgFlag ~= 0 )
+%                         resvec = abs(e*tempK - K(ioffset+1:ioffset+NVAR));
+%                         scalar = OPTIONS.AbsTol + OPTIONS.RelTol.*abs(K(ioffset+1:ioffset+NVAR));
+%                         if (norm(resvec./scalar) > sqrt(NVAR))
+%                             singCount = singCount + 1;
+%                             switch(bicgFlag)
+%                                 case 1
+%                                     warning('BICG: iterated MAXIT times but did not converge');
+%                                     break;
+%                                 case 2
+%                                     warning('BICG: preconditioner M was ill-conditioned');
+%                                     break;
+%                                 case 3
+%                                     warning('BICG: stagnated (two consecutive iterates were the same)');
+%                                     break;
+%                                 case 4
+%                                     warning('BICG: One of the scalar quantities became too small or too large to continue computing.');
+%                                     break;
+%                             end
+%                         else
+%                             bicgFlag = 0;
+%                         end
+%                     end
+%                     K(ioffset+1:ioffset+NVAR) = tempK;
+
+                else
+                    [tempK, ISTATUS, gmresFlag, singFlag] = MatrixFreeSolve(e, K(ioffset+1:ioffset+NVAR), OPTIONS, ISTATUS);
+                    if singFlag
+                        singCount = singCount + 1;
+                    end
+                    K(ioffset+1:ioffset+NVAR) = tempK;
+                end
                 
             end % stages
             
+            if( gmresFlag ~= 0 )
+                H = 0.5*H;
+                continue;
+            end
+            
```

Comment:
Is there something better that we can do instead of halving the step-size if a gmres iteration fails? Maybe there should be a configurable scaling factor or other such thing?